botbot_plugins/tests/test_vote.py

```diff
@@ -0,0 +1,247 @@
+    app.respond("!vote app")
+    assert app.respond("!countvotes") == ["[app(+1, -0): repl_user] [firewall(+0, -0): ] [port(+0, -0): ] [\\0: ]"]
+
+    # Closest match
+    app.respond("-app is not cool")
+    assert app.respond("!countvotes") == ["[app(+0, -1): -repl_user] [firewall(+0, -0): ] [port(+0, -0): ] [\\0: ]"]
+    app.respond("+port is a good one")
+    assert app.respond("!countvotes") == ["[app(+0, -1): -repl_user] [firewall(+0, -0): ] [port(+1, -0): repl_user] [\\0: ]"]
+    app.respond("+app is kinda cool")
+    assert app.respond("!countvotes") == ["[app(+1, -0): repl_user] [firewall(+0, -0): ] [port(+1, -0): repl_user] [\\0: ]"]
+
+    # Multiple users
+    app.respond("-app", User="jesus")
+    assert app.respond("!countvotes") == ["[app(+1, -1): repl_user; -jesus] [firewall(+0, -0): ] [port(+1, -0): repl_user] [\\0: ]"]
+    app.respond("+port", User="eladio")
+    assert app.respond("!countvotes") == ["[app(+1, -1): repl_user; -jesus] [firewall(+0, -0): ] [port(+2, -0): repl_user, eladio] [\\0: ]"]
+    app.respond("+app", User="eladio")
+    assert app.respond("!countvotes") == ["[app(+2, -1): repl_user, eladio; -jesus] [firewall(+0, -0): ] [port(+2, -0): repl_user, eladio] [\\0: ]"]
+    app.respond("\\1", User="jesus")
+    assert app.respond("!countvotes") == ["[app(+2, -0): repl_user, eladio] [firewall(+0, -0): ] [port(+2, -0): repl_user, eladio] [\\1: jesus]"]
+
+    # Invalid option
+    assert app.respond("+invalid") == []
+    assert app.respond("!vote +invalid") == [u"“invalid” is not a valid option."]
+
+    assert app.respond("!endvote") == ["Voting has ended."]
+
+    # With name
+    assert app.respond("!startvote is jesus real? [port, firewall, app]") == [u"Voting has started for proposal “is jesus real?”."]
+    assert app.respond("!endvote") == [u"Voting has ended for proposal “is jesus real?”."]
+
+    # Closest match needs to be greedy
+    app.respond("!startvote [hello, hello world]")
+    app.respond("+hello")
+    assert app.respond("!countvotes") == ["[hello(+1, -0): repl_user] [hello world(+0, -0): ] [\\0: ]"]
+    app.respond("-hello world")
+    assert app.respond("!countvotes") == ["[hello(+1, -0): repl_user] [hello world(+0, -1): -repl_user] [\\0: ]"]
+    app.respond("+hello\tworld")
+    assert app.respond("!countvotes") == ["[hello(+1, -0): repl_user] [hello world(+1, -0): repl_user] [\\0: ]"]
+    app.respond("!endvote")
+
+    # Duplicate options
+    app.respond("!startvote [hello, hello, world, world]")
+    app.respond("+hello")
+    app.respond("-world")
+    assert app.respond("!countvotes") == ["[hello(+1, -0): repl_user] [world(+0, -1): -repl_user] [\\0: ]"]
+    app.respond("!endvote")
+
+    # With explicitly empty options
+    assert app.respond("!startvote []") == ["Voting has started."]
+    app.respond("+1")
+    assert app.respond("!countvotes") == ["[+1: repl_user] [-0: ] [\\0: ]"]
```

Comment:
Duplicate assertion code. Make a function from this maybe?

Reply:
For me, the advantage with this is that it's clear exactly what's being passed into the plugin, without having to consult an external method. Besides, this method would be kind of hard to generalize,  although there are some ways. Lemme see if I can do something.